### PR TITLE
Add oci cloud controller

### PR DESCRIPTION
**Pull request by garthy**

### CODE REVIEW

File: instances/k8smaster/scripts/kubelet.service [hunk 1]
```diff
@@ -8,6 +8,7 @@ ExecStart=/usr/bin/kubelet \
   --kubeconfig=http://127.0.0.1:8080 \
   --pod-manifest-path=/etc/kubernetes/manifests \
   --hostname-override=__FQDN_HOSTNAME__ \
+  --cloud-provider=external \
```
Comment (prydie) [inline instances/k8smaster/scripts/kubelet.service]: Should also set `--provider-id` here so that we're not relying on the CCM to resolve the node name to OCID.

File: instances/k8sworker/scripts/kubelet.service [hunk 2]
```diff
@@ -21,6 +21,7 @@ ExecStart=/usr/bin/kubelet \
   --tls-private-key-file=/etc/kubernetes/ssl/apiserver-key.pem \
   --feature-gates=ExperimentalCriticalPodAnnotation=true \
   --eviction-hard=memory.available<500Mi,nodefs.available<2Gi,imagefs.available<2Gi \
+  --cloud-provider=external \
```
Comment (prydie) [inline instances/k8sworker/scripts/kubelet.service]: Should also set `--provider-id` here so that we're not relying on the CCM to resolve the node name to OCID.

File: kubernetes/oci-cloud-controller/user.tf [hunk 3]
```diff
@@ -0,0 +1,44 @@
+resource "tls_private_key" "cloud_controller_user_key" {
+  algorithm   = "RSA"
+  rsa_bits = 2048
+}
+
+resource "oci_identity_group" "cloud_controller_group" {
+    name = "${var.label_prefix}cloud_controller_group"
+    description = "Terraform created group for ocii cloud controller"
```
Comment (prydie) [inline kubernetes/oci-cloud-controller/user.tf]: `s/ocii cloud controller/OCI Cloud controller manager/`

File: kubernetes/oci-cloud-controller/user.tf [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+}
+
+resource "oci_identity_user" "cloud_controller_user" {
+  name = "${var.label_prefix}cloud_controller_user"
+  description = "Terraform created user for ocii cloud controller"
```
Comment (prydie) [inline kubernetes/oci-cloud-controller/user.tf]: `s/ocii cloud controller/OCI Cloud controller manager/`

File: kubernetes/oci-cloud-controller/user.tf [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+  statements = [
+    "Allow group id ${oci_identity_group.cloud_controller_group.id} to manage load-balancers in tenancy", #Restrict to compartment?
+    "Allow group id ${oci_identity_group.cloud_controller_group.id} to use security-lists in tenancy",
+    "Allow group id ${oci_identity_group.cloud_controller_group.id} to read instances in tenancy",
+    "Allow group id ${oci_identity_group.cloud_controller_group.id} to read subnets in tenancy",
+    "Allow group id ${oci_identity_group.cloud_controller_group.id} to read vnics in tenancy",
+    "Allow group id ${oci_identity_group.cloud_controller_group.id} to read vnic-attachments in tenancy",
+  ]
+}
+
```
Comment (prydie) [inline kubernetes/oci-cloud-controller/user.tf]: Trailing `\n`s

File: k8s-oci.tf [hunk 6]
```diff
@@ -42,6 +42,16 @@ module "vcn" {
   worker_nodeport_ingress                 = "${var.worker_nodeport_ingress}"
 }
 
+module "oci-cloud-controller" {
```
Comment (prydie) [inline k8s-oci.tf]: `s/oci-cloud-controller/oci-cloud-controller-manager/`?

File: instances/k8smaster/scripts/kubelet.service (hunk 1 above)
Comment (garthy) [reply instances/k8smaster/scripts/kubelet.service]: OK I looked at the docs for the cloud controller and it said that I didn't need this for 1.7.6+

File: kubernetes/oci-cloud-controller/user.tf (hunk 3 above)
Comment (garthy) [reply kubernetes/oci-cloud-controller/user.tf]: Urgh thanks.

File: kubernetes/oci-cloud-controller/datasources.tf [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+    user= "${oci_identity_user.cloud_controller_user.id}"
+    key = "${jsonencode(tls_private_key.cloud_controller_user_key.private_key_pem)}"
+    
+    subnet1 = "${var.subnet1}"
+    subnet2 = "${var.subnet1}"
+  }
+}
+
+data "template_file" "cloud-provider-yaml" {
+  template = "${file("${path.module}/cloud-provider.yaml")}"
```
Comment (jlamillan) [inline kubernetes/oci-cloud-controller/datasources.tf]: Where is this file?

File: kubernetes/oci-cloud-controller/user.tf (hunk 5 above)
Comment (jlamillan) [reply kubernetes/oci-cloud-controller/user.tf]: My current user only has compartment level permissions, so running `terraform apply` fails with a `Status: 404; Code: NotAuthorizedOrNotFound`.

What do you think about _only_ trying to generae a new user and group, etc. if the following variables are _not_ set:

```
cloud_controller_user_ocid
cloud_controller_user_fingerprint
cloud_controller_user_private_key_path
```

Otherwise, the user running Terraform now has to have a root level permissions instead of just compartment level permissions as they do today.

**Comment by jlamillan** [comment]
Does this change _require_ Kubernetes 1.7.10?

File: instances/k8smaster/scripts/kubelet.service (hunk 1 above)
Comment (garthy) [reply instances/k8smaster/scripts/kubelet.service]: I'm going to leave this a seperate MR.

File: kubernetes/oci-cloud-controller/datasources.tf (hunk 7 above)
Comment (garthy) [reply kubernetes/oci-cloud-controller/datasources.tf]: Missing!

**Comment by garthy** [comment]
This change requires 1.7.6+ (Maybe 1.7.2+ if I make the provider-id changes Andrew mention)

File: kubernetes/oci-cloud-controller/datasources.tf [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+    region      = "${var.region}"
+    compartment = "${var.compartment_ocid}"
+    
+    # if cloud_controller_user_ocid is empty string then generate a user/group/policy and use it here
+    fingerprint = "${var.cloud_controller_user_ocid=="" ? join("", oci_identity_api_key.cloud_controller_key_assoc.*.fingerprint) : var.cloud_controller_user_fingerprint }"
+    user        = "${var.cloud_controller_user_ocid=="" ? join("", oci_identity_user.cloud_controller_user.*.id) : var.cloud_controller_user_ocid }"
+    key         = "${var.cloud_controller_user_ocid=="" ? jsonencode(join("", tls_private_key.cloud_controller_user_key.*.private_key_pem)) : jsonencode(file(var.cloud_controller_user_private_key_path))}"
+    
+    subnet1 = "${var.subnet1}"
+    subnet2 = "${var.subnet1}"
```
Comment (jlamillan) [inline kubernetes/oci-cloud-controller/datasources.tf]: Typo - should read `subnet2 = "${var.subnet2}"`

File: k8s-oci.tf [hunk 9]
```diff
@@ -42,6 +42,21 @@ module "vcn" {
...
+  label_prefix                           = "${var.label_prefix}"
+  compartment_ocid                       = "${var.compartment_ocid}"
+  tenancy                                = "${var.tenancy_ocid}"
+  region                                 = "${var.region}"
+
+  cloud_controller_user_ocid             = "${var.cloud_controller_user_ocid}"
+  cloud_controller_user_fingerprint      = "${var.cloud_controller_user_fingerprint}"
+  cloud_controller_user_private_key_path = "${var.cloud_controller_user_private_key_path}"
+  
+  subnet1                                = "${module.vcn.etcd_subnet_ad1_id}"
```
Comment (jlamillan) [inline k8s-oci.tf]: This should be the worker subnet here right? 

```
subnet1                                = "${module.vcn.k8worker_subnet_ad1_id}"
subnet2                                = "${module.vcn.k8worker_subnet_ad2_id}"
```

**Comment by jlamillan** [comment]
On one hand, I like the idea of a dedicated user/group that _only_ has permissions for the load-balancer(s) since the OCIDs will live in the k8s cluster as a `secret`. On the other hand, it feels counter intuitive that even though the Installer only creates artifacts in a single compartment, the user creating the cluster suddenly has to have _tenant level_ permissions (to create the user, group, and policy). From what we've seen, most of the people spinning up clusters only have compartment level permissions in the first place.

I think the most backwards compatible approach that still preserves your handy code for creating a custom user, group, and policy would be:

1. If `cloud_controller*` input variables are *not* set, fall back to using `user_ocid` (i.e. the user that is doing the `terraform apply` - this would probably be the path for most installs.
2. If `cloud_controller*` input variables *are* set, use those for the cloud controller user and group.

For (2), if `kubernetes/oci-cloud-controller/user.tf` was moved `identity/` and unhooked from `k8s-oci.tf`, it wouldn't automatically be called as a part of cluster creation. The user could still chose to execute a `terraform apply` directly from `identity/` to create `oci_identity_user`, `oci_identity_group`, `oci_identity_user_group_membership` and then output the  `cloud_controller*` values, which could be used as input variables.

Basically, `identity/` would be an _optional_ pre-cluster creation step for potentially more advanced / users, which we could doc in README.

**Comment by jlamillan** [comment]
Another reason why it makes sense for separate `terraform apply` is that it will error out if you try and create a user / group / etc. outside your "home region". So, the regions for the cluster and any identity resources will often need to be in different.

**Comment by jlamillan** [comment]
@garthy [here](https://github.com/oracle/terraform-kubernetes-installer/tree/jlamillan/gb-oci-cloud-controller) is the updated branch with the changes described above.

File: kubernetes/oci-cloud-controller/cloud-provider.yaml [hunk 10]
```diff
@@ -0,0 +1,12 @@
+{
```
Comment (owainlewis) [inline kubernetes/oci-cloud-controller/cloud-provider.yaml]: Filename should be .json

File: kubernetes/oci-cloud-controller/datasources.tf (hunk 7 above)
Comment (owainlewis) [reply kubernetes/oci-cloud-controller/datasources.tf]: After you update the file above => cloud-provider.json

File: kubernetes/oci-cloud-controller/user.tf [hunk 11]
```diff
@@ -0,0 +1,46 @@
+resource "tls_private_key" "cloud_controller_user_key" {
+  count = "${var.cloud_controller_user_ocid==""? 1 : 0 }"
+  algorithm   = "RSA"
+  rsa_bits = 2048
+}
+
+resource "oci_identity_group" "cloud_controller_group" {
+  count = "${var.cloud_controller_user_ocid==""? 1 : 0 }"
+  name = "${var.label_prefix}cloud_controller_group"
+  description = "Terraform created group for OCI Cloud controller manager"
```
Comment (owainlewis) [inline kubernetes/oci-cloud-controller/user.tf]: Consistent capitalisation 

Terraform created group for OCI Cloud Controller manager

File: kubernetes/oci-cloud-controller/user.tf [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+}
+
+resource "oci_identity_user" "cloud_controller_user" {
+  count = "${var.cloud_controller_user_ocid==""? 1 : 0 }"
+  name = "${var.label_prefix}cloud_controller_user"
+  description = "Terraform created user for OCI Cloud controller manager"
```
Comment (owainlewis) [inline kubernetes/oci-cloud-controller/user.tf]: Terraform created group for OCI Cloud Controller Manager

File: instances/k8smaster/scripts/kubelet.service (hunk 1 above)
Comment (jlamillan) [reply instances/k8smaster/scripts/kubelet.service]: We apparently need to move the default back to 1.7.4. So, I'll add the `provider-id` as well to my changes.

**Comment by jlamillan** [comment]
Tracking this integration work in [this](https://github.com/oracle/terraform-kubernetes-installer/pull/100) PR.

File: kubernetes/oci-cloud-controller/datasources.tf (hunk 8 above)
Comment (garthy) [reply kubernetes/oci-cloud-controller/datasources.tf]: This is actually wrong and I've misunderstood the CCM LB requirements. I need to create some subnets for the load balancers to live on.

**Comment by garthy** [comment]
So I don't like 1) The fall back as it is a bit of a security issue. Any admin kube users will have access to oci creds that can do some pretty bad stuff. The CCM policy restrictions mitigate this somewhat.
